hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -198,111 +214,135 @@ public Object convert(Object value) {
-            if (field.getType().isCustomType()) {
-                if (!typeMap.containsKey(field.type.objectTypeName)) {
-                    collectCustomTypes(field.type, typeMap);
+            Map<String, QueryDataType> typeMap = new HashMap<>();
+            collectCustomTypes(type, typeMap);
+
+            out.writeInt(typeMap.size());
+            for (QueryDataType nestedType : typeMap.values()) {
+                writeObjectTypeMetadata(nestedType, out);
+                out.writeInt(nestedType.getObjectFields().size());
+                for (QueryDataTypeField field : nestedType.getObjectFields()) {
+                    out.writeString(field.name);
+                    out.writeInt(field.type.converter.getId());
+                    writeObjectTypeMetadata(field.type, out);
                 }
             }
         }
-    }
 
-    @Override
-    public void readData(ObjectDataInput in) throws IOException {
-        converter = Converters.getConverter(in.readInt());
-        if (converter.getTypeFamily() != QueryDataTypeFamily.OBJECT) {
-            return;
+        private static void writeObjectTypeMetadata(QueryDataType type, ObjectDataOutput out) throws IOException {
+            out.writeInt(type.objectTypeKind.value());
+            out.writeString(type.objectTypeName);
+            out.writeString(type.objectTypeMetadata);
         }
 
-        readObjectTypeMetadata(this, in);
-        if (!isCustomType()) {
-            return;
+        private void collectCustomTypes(QueryDataType type, Map<String, QueryDataType> typeMap) {
+            typeMap.put(type.objectTypeName, type);
+
+            for (QueryDataTypeField field : type.objectFields) {
+                if (field.getType().isCustomType()) {
+                    if (!typeMap.containsKey(field.type.objectTypeName)) {
+                        collectCustomTypes(field.type, typeMap);
+                    }
+                }
+            }
         }
 
-        Map<String, QueryDataType> typeMap = new HashMap<>();
-        typeMap.put(objectTypeName, this);
-
-        int typeMapSize = in.readInt();
-        for (int i = 0; i < typeMapSize; i++) {
-            QueryDataType type = readNestedType(OBJECT.getConverter(), in, typeMap);
-            int fields = in.readInt();
-            for (int j = 0; j < fields; j++) {
-                String fieldName = in.readString();
-                Converter converter = Converters.getConverter(in.readInt());
-                QueryDataType nestedType = readNestedType(converter, in, typeMap);
-                type.getObjectFields().add(new QueryDataTypeField(fieldName, nestedType));
+        @Nonnull @Override
+        public QueryDataType read(ObjectDataInput in) throws IOException {
+            Converter converter = Converters.getConverter(in.readInt());
+            if (converter.getTypeFamily() != QueryDataTypeFamily.OBJECT) {
+                return TYPES_BY_CONVERTER.get(converter);
```

Comment:
lookup by converter id in `Int2ObjectHashMap` should be more efficient

Reply:
or even better in plain array: `Converters.CONVERTERS`

Reply:
I didn't want to depend on the implementation details of `Converters` to make such an optimization, but since it is part of the serialized format, `Converter` IDs cannot change. Thus, we can use a similar array called `TYPES`, which is in the same order as `Converter`s. I also exposed it via `values()` method to make `QueryDataType` even more enum-like. (5b988d41)

Reply:
✔️